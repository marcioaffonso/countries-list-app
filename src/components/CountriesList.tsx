import React, { useDeferredValue } from 'react';
import { View, Text, FlatList, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';

export default function CountriesList({ countries, onRefresh, refreshing }) {
  const deferredCustomers = useDeferredValue(countries);
  return (
    <FlatList
      data={deferredCustomers}
      renderItem={({ item }) => (
        <View style={styles.countryCard}>
          <View style={styles.nameContainer}>
            <Text style={styles.name}>{item.name.common}</Text>
            <Text style={styles.flag}>{item.flag}</Text>
          </View>
          <View style={styles.details}>
            <Text style={styles.detailsText}>üåé Name: { item.name.official }</Text>
            <Text style={styles.detailsText}>üè† Capital: { item.capital.join(', ') }</Text>
            <Text style={styles.detailsText}>
              üíµ Currency: { Object.keys(item.currencies).map((key) => {
                const currency = item.currencies[key];
                return `${key} (${currency.symbol}) - ${currency.name}`;
              }).join(', ') }
            </Text>
          </View>
        </View>
      )}
      ListEmptyComponent={() => (
        <View style={styles.countryCard}>
          <Text style={styles.detailsText}>No countries matching the filter criteria.</Text>
        </View>
      )}
      onRefresh={onRefresh}
      refreshing={refreshing}
    />
  );
};

CountriesList.propTypes = {
  countries: PropTypes.array,
  onRefresh: PropTypes.func,
  refreshing: PropTypes.bool
};

const styles = StyleSheet.create({
  countryCard: {
    backgroundColor: '#fff',
    borderRadius: 16,
    borderWidth: 2,
    margin: 10,
    padding: 16
  },
  nameContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: 20
  },
  name: {
    fontSize: 30,
    fontWeight: 'bold',
    flexWrap: 'wrap'
  },
  flag: {
    fontSize: 30
  },
  details: {
    marginBottom: 10
  },
  detailsText: {
    fontSize: 18
  },
});
