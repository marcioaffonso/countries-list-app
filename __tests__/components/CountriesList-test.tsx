import React from 'react';
import renderer from 'react-test-renderer';
import { expect, test } from '@jest/globals';
import CountriesList from '../../src/components/CountriesList';
import { countries } from '../mocks/countries';

test('renders empty countries list correctly', () => {
  const tree = renderer.create(<CountriesList countries={[]} />).toJSON;
  expect(tree).toMatchSnapshot();
});

test('renders filled countries list correctly', () => {
  const tree = renderer.create(<CountriesList countries={countries} />);
  expect(tree).toMatchSnapshot();
});
