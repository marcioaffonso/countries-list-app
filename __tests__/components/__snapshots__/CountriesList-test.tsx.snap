// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty countries list correctly 1`] = `[Function]`;

exports[`renders filled countries list correctly 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  data={
    [
      {
        "capital": [
          "Capital1",
        ],
        "currencies": {
          "C1": {
            "name": "currenycy1",
            "symbol": "$",
          },
        },
        "flag": "🇧🇷",
        "name": {
          "common": "Country1",
          "official": "Official1",
        },
      },
      {
        "capital": [
          "Capital2",
        ],
        "currencies": {
          "C2": {
            "name": "currenycy2",
            "symbol": "$",
          },
        },
        "flag": "🇺🇸",
        "name": {
          "common": "Country2",
          "official": "Official2",
        },
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "backgroundColor": "#fff",
            "borderRadius": 16,
            "borderWidth": 2,
            "margin": 10,
            "padding": 16,
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 20,
            }
          }
        >
          <Text
            style={
              {
                "flexWrap": "wrap",
                "fontSize": 30,
                "fontWeight": "bold",
              }
            }
          >
            Country1
          </Text>
          <Text
            style={
              {
                "fontSize": 30,
              }
            }
          >
            🇧🇷
          </Text>
        </View>
        <View
          style={
            {
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              {
                "fontSize": 18,
              }
            }
          >
            🌎 Name: 
            Official1
          </Text>
          <Text
            style={
              {
                "fontSize": 18,
              }
            }
          >
            🏠 Capital: 
            Capital1
          </Text>
          <Text
            style={
              {
                "fontSize": 18,
              }
            }
          >
            💵 Currency: 
            C1 ($) - currenycy1
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "backgroundColor": "#fff",
            "borderRadius": 16,
            "borderWidth": 2,
            "margin": 10,
            "padding": 16,
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 20,
            }
          }
        >
          <Text
            style={
              {
                "flexWrap": "wrap",
                "fontSize": 30,
                "fontWeight": "bold",
              }
            }
          >
            Country2
          </Text>
          <Text
            style={
              {
                "fontSize": 30,
              }
            }
          >
            🇺🇸
          </Text>
        </View>
        <View
          style={
            {
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              {
                "fontSize": 18,
              }
            }
          >
            🌎 Name: 
            Official2
          </Text>
          <Text
            style={
              {
                "fontSize": 18,
              }
            }
          >
            🏠 Capital: 
            Capital2
          </Text>
          <Text
            style={
              {
                "fontSize": 18,
              }
            }
          >
            💵 Currency: 
            C2 ($) - currenycy2
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
